internal/ethapi/api_extra.go

```diff
@@ -113,12 +114,14 @@ func (s *BlockChainAPI) stateQueryBlockNumberAllowed(blockNumOrHash rpc.BlockNum
 	var number uint64
 	if blockNumOrHash.BlockNumber != nil {
 		number = uint64(blockNumOrHash.BlockNumber.Int64())
-	} else {
-		block, err := s.b.BlockByNumberOrHash(context.Background(), blockNumOrHash)
-		if err != nil {
+	} else if blockHash, ok := blockNumOrHash.Hash(); ok {
+		block, err := s.b.BlockByHash(context.Background(), blockHash)
+		if block == nil || err != nil {
```

Comment:
if the block is nil, you cannot wrap the error below really (as in it won't panic, but it will wrap oddly with the nil error)

```suggestion
		if block == nil {
			return fmt.Errorf("block from hash doesn't exist")
		} else if err != nil {
```